src/AppInstallerRepositoryCore/ARPCorrelation.cpp

```diff
@@ -0,0 +1,442 @@
+            bestMatchingScore = std::max(bestMatchingScore, matchingScore);
+        }
+
+        return bestMatchingScore;
+    }
+
+    std::optional<ARPEntry> ARPCorrelationAlgorithm::GetBestMatchForManifest(
+        const Manifest::Manifest& manifest,
+        const std::vector<ARPEntry>& arpEntries) const
+    {
+        AICLI_LOG(Repo, Verbose, << "Looking for best match in ARP for manifest " << manifest.Id);
+
+        std::optional<ARPEntry> bestMatch;
+        double bestScore = std::numeric_limits<double>::lowest();
+
+        for (const auto& arpEntry : arpEntries)
+        {
+            auto score = GetMatchingScore(manifest, arpEntry);
+            AICLI_LOG(Repo, Verbose, << "Match score for " << arpEntry.Entry->GetProperty(PackageVersionProperty::Id) << ": " << score);
+
+            if (score < MatchingThreshold)
+            {
+                AICLI_LOG(Repo, Verbose, << "Score is lower than threshold");
+                continue;
+            }
+
+            if (!bestMatch || bestScore < score)
+            {
+                bestMatch = arpEntry;
+                bestScore = score;
+            }
+        }
+
+        if (bestMatch)
+        {
+            AICLI_LOG(Repo, Verbose, << "Best match is " << bestMatch->Entry->GetProperty(PackageVersionProperty::Id));
+        }
+        else
+        {
+            AICLI_LOG(Repo, Verbose, << "No ARP entry had a correlation score surpassing the required threshold");
+        }
+
+        return bestMatch;
+    }
+
+    const ARPCorrelationAlgorithm& ARPCorrelationAlgorithm::Instance()
+    {
+        return InstanceInternal();
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    void ARPCorrelationAlgorithm::OverrideInstance(ARPCorrelationAlgorithm* algorithmOverride)
+    {
+        InstanceInternal(algorithmOverride);
+    }
+
+    void ARPCorrelationAlgorithm::ResetInstance()
+    {
+        InstanceInternal(nullptr);
+    }
+#endif
+
+    double NormalizedNameAndPublisherCorrelationMeasure::GetMatchingScore(std::string_view packageName, std::string_view packagePublisher, std::string_view arpName, std::string_view arpPublisher) const
+    {
+        NameNormalizer normer(NormalizationVersion::Initial);
+
+        auto packageNormalizedName = normer.Normalize(packageName, packagePublisher);
+        auto arpNormalizedName = normer.Normalize(arpName, arpPublisher);
+
+        if (Utility::CaseInsensitiveEquals(arpNormalizedName.Name(), packageNormalizedName.Name()) &&
+            Utility::CaseInsensitiveEquals(arpNormalizedName.Publisher(), packageNormalizedName.Publisher()))
+        {
+            return 1;
+        }
+        else
+        {
+            return 0;
+        }
+    }
+
+    double EditDistanceNormalizedNameAndPublisherCorrelationMeasure::GetMatchingScore(std::string_view packageName, std::string_view packagePublisher, std::string_view arpName, std::string_view arpPublisher) const
+    {
+        NameNormalizer normer(NormalizationVersion::Initial);

```

Comment:
I changed it to have an object initialized with the manifest. It did make a noticeable difference in the time it took. :)
Since I was doing that, I also changed it to use the `IARPMatchConfidenceAlgorithm` interface you had suggested earlier.

> On top of that, for now the name and publisher values coming out of the ARP snapshot are already normalized which can save some time.
I tried not normalizing them, but then the success rate dropped significantly. Seems like they are not normalized? So I'm keeping that normalization, although it did make a difference in the time removing that normalization :(